panic: sync: WaitGroup is reused before previous Wait has returned

### Background

I haven't updated in a while (since 0.4.1-beta commit= bdf3f4d7753ce950bd14ed56ad96c12097f1b4c9 ) and just pull from master an ~hour ago. The node is pretty loaded with 170+ mainnet channels, although the server is pretty beefy.
Anyway, it seems after a channel of a rejected commitment fails, the htlcManager exits there's concurrency panic, but I'm clueless :stuck_out_tongue: 

```
2018-05-26 01:13:40.513 [INF] PEER: unable to read message from 217.30.71.110:9735: EOF
2018-05-26 01:13:40.513 [ERR] HSWC: ChannelLink(518205:244:0) Failing link: ChannelPoint(4c858dc50040ebf316f51db84eff632d2d9c6a071b155213fa0ca36280933041:0): received error from peer: rejected commitment: commit_height=281, invalid_commit
_sig=304502210081f3b5c7f3a2ccab5ead863de980f450b8e35e6ab318108d2286eb1d3755d7c1022040c170d3bbd43317123a6e677c9fbef2f683c392dc1e79324bb02a1fa79116e8, commit_tx=02000000014130938062a30cfa1352151b076a9c2d2d63ff4eb81df516f3eb4000c58d854c00000
0000041ed928002d855010000000000220020a07f6323a4c5beacbba012d9376048034addb9e581b315283c7bb6e2e067bfc10ab1010000000000160014dc85ea82d153436234f3fb53c414a91d897b508ca5c78220, sig_hash=d6cf176dd3411aeb7c899feb7842d3af184aa6a32245bb1dbb275914
86963c55
2018-05-26 01:13:40.513 [ERR] HSWC: ChannelLink(518205:244:0) link failed, exiting htlcManager
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 589930 [running]:
sync.(*WaitGroup).Wait(0xc422d98158)
        /usr/lib/go-1.10/src/sync/waitgroup.go:131 +0xbb
main.(*peer).WaitForDisconnect(0xc422d88000)
        /root/gocode/src/github.com/lightningnetwork/lnd/peer.go:557 +0x33
main.(*server).peerTerminationWatcher(0xc420556a00, 0xc422d88000)
        /root/gocode/src/github.com/lightningnetwork/lnd/server.go:1291 +0x79
created by main.(*server).addPeer
        /root/gocode/src/github.com/lightningnetwork/lnd/server.go:1780 +0x183
```
Crashes at that point, after I start it again, a minute or two later I get the same error:
```
2018-05-26 01:18:48.780 [INF] PEER: unable to read message from 217.30.71.110:9735: EOF
2018-05-26 01:18:48.780 [ERR] HSWC: ChannelLink(518205:244:0) Failing link: ChannelPoint(4c858dc50040ebf316f51db84eff632d2d9c6a071b155213fa0ca36280933041:0): received error from peer: rejected commitment: commit_height=281, invalid_commit_sig=304502210081f3b5c7f3a2ccab5ead863de980f450b8e35e6ab318108d2286eb1d3755d7c1022040c170d3bbd43317123a6e677c9fbef2f683c392dc1e79324bb02a1fa79116e8, commit_tx=02000000014130938062a30cfa1352151b076a9c2d2d63ff4eb81df516f3eb4000c58d854c000000000041ed928002d855010000000000220020a07f6323a4c5beacbba012d9376048034addb9e581b315283c7bb6e2e067bfc10ab1010000000000160014dc85ea82d153436234f3fb53c414a91d897b508ca5c78220, sig_hash=d6cf176dd3411aeb7c899feb7842d3af184aa6a32245bb1dbb27591486963c55
2018-05-26 01:18:48.780 [ERR] HSWC: ChannelLink(518205:244:0) link failed, exiting htlcManager
panic: sync: WaitGroup is reused before previous Wait has returned

goroutine 62883 [running]:
sync.(*WaitGroup).Wait(0xc45f732158)
        /usr/lib/go-1.10/src/sync/waitgroup.go:131 +0xbb
main.(*peer).WaitForDisconnect(0xc45f722000)
        /root/gocode/src/github.com/lightningnetwork/lnd/peer.go:557 +0x33
main.(*server).peerTerminationWatcher(0xc4200bb2c0, 0xc45f722000)
        /root/gocode/src/github.com/lightningnetwork/lnd/server.go:1291 +0x79
created by main.(*server).addPeer
        /root/gocode/src/github.com/lightningnetwork/lnd/server.go:1780 +0x183
```

### Your environment

* lncli version 0.4.1 commit= dd2859dc61a11f878da82ed875228cc62cd9d47e (fresh, ~1 hour ago)
* Ubuntu 16.04 LTS - Linux 4.4.0-45-generic 66-Ubuntu SMP Wed Oct 19 14:12:37 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
* `bitcoind` v0.16
* Pretty moderate machine
* The node has 170+ channels (mainnet)

### Steps to reproduce

Tell us how to reproduce this issue. Please provide stacktraces and links to code in question.

### Expected behaviour

Handle rejected commitments on failed links?

### Actual behaviour

Error/concurrency panic?

--- a/peer.go
+++ b/peer.go
@@ -438,14 +438,23 @@ func (p *peer) addLink(chanPoint *wire.OutPoint,
 		shortChanID lnwire.ShortChannelID,
 		linkErr htlcswitch.LinkFailureError) {
 
+		select {
+		// If the server is already exiting, then none of the actions
+		// below can finish exiting, so we'll exit early as well.
+		case <-p.server.quit:
+			return
+
+		default:
+		}
+
 		// The link has notified us about a failure. We launch a go
 		// routine to stop the link, disconnect the peer and optionally
 		// force close the channel. We must launch a goroutine since we
 		// must let OnChannelFailure return in order for the link to
 		// completely stop in the call to RemoveLink.
-		p.wg.Add(1)
+		p.server.wg.Add(1)
 		go func() {
-			defer p.wg.Done()
+			defer p.server.wg.Done()
 
 			// We begin by removing the link from the switch, such
 			// that it won't be attempted used for any more

